File: Leetcode/652/652.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

struct Tuple {
    int val, l, r;
    Tuple(int val, int l, int r) : val(val), l(l), r(r) {}
    Tuple() = default;
    bool operator == (const Tuple &node) const {
        return val == node.val && l == node.l && r == node.r;
    }
};

struct myHashFunc {
    size_t operator() (const Tuple& node) const {
        size_t h = 0;
        h ^= hash<int>{}(node.val) + 0x9e3779b9 + (h << 6) + (h >> 2);
        h ^= hash<int>{}(node.l) + 0x9e3779b9 + (h << 6) + (h >> 2);
        h ^= hash<int>{}(node.r) + 0x9e3779b9 + (h << 6) + (h >> 2);
        return h;
    }
};

class Solution {
public:
    vector<TreeNode*> ans;
    unordered_map<Tuple, int, myHashFunc> dp;   // tuple to its index
    unordered_map<int,int> freq;    // tuple's index to its oucurrences
    int globalId;
    int dfs(TreeNode* node) {
        if (node) {
            auto k = Tuple(node->val, dfs(node->left), dfs(node->right));
            auto iter = dp.find(k);
            if (iter == dp.end()) {
                dp[k] = globalId;
                ++freq[globalId];
                ++globalId;
            }
            else {
                ++freq[dp[k]];
            }
            if (freq[dp[k]] == 2) {
                ans.emplace_back(node);
            }
            return dp[k];
        }
        else {
            return -1;
        }
    }
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        vector<TreeNode*>().swap(ans);
        dp.clear();
        freq.clear();
        globalId = 0;
        dfs(root);
        return ans;
    }
};

File: Leetcode/652/652.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type Tuple struct {
	val, l, r int
}

var (
	dp map[Tuple]int
	freq map[int]int
	id int
	ans []*TreeNode
)

func dfs(node *TreeNode) int {
	if node != nil {
		k := Tuple{
			val: node.Val,
			l:   dfs(node.Left),
			r:   dfs(node.Right),
		}
		val, ok := dp[k]
		if !ok {
			dp[k] = id
			freq[id]++
			id++
		} else {
			freq[val]++
			if freq[val] == 2 {
				ans = append(ans, node)
			}
		}
		return dp[k]
	}
	return -1
}

func findDuplicateSubtrees(root *TreeNode) []*TreeNode {
	dp = nil
	dp = make(map[Tuple]int)
	freq = nil
	freq = make(map[int]int)
	id = 0
	ans = nil
	_ = dfs(root)
	return ans
}